README.md
# Music is mid playlist bot

Monitor the Music is Mid Discord server for Spotify links and add them to a playlist.

## Development

Run the project's tests:

    $ clojure -T:build test

Run the project's CI pipeline and build an uberjar:

    $ clojure -T:build ci

Run that uberjar:

    $ java -jar target/myapp-0.1.0-SNAPSHOT.jar


## License

This code is currently not provided under any license.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 0.1.1 - 2023-08-17
Fixed `get-fresh-token` logic, which was not recognising expired tokens.

## 0.1.0 - 2023-08-17
First version

test/mim/spotify_test.clj
(ns mim.spotify-test
  (:require [clojure.test :refer :all]
            [mim.spotify :refer :all]))

(deftest url-extractors
  (is (extract-tracks
       (str "Lorem https://open.spotify.com/track/61NudzDPuUUH9NuhKMDgRD?"
            "si=ec51a9fa907b4ed0 ipsum sit dolor https://open.spotify.com"
            "/track/2aQ8NyRJcsQFgb8jwUil90?si=c0671cf6eeb04970\n"
            "ipsum"))
      ["spotify:track:61NudzDPuUUH9NuhKMDgRD"
       "spotify:track:2aQ8NyRJcsQFgb8jwUil90"])
  (is
   (extract-playlists
    "lorem https://open.spotify.com/playlist/37i9dQZF1DZ06evO2HVeMk?si=5f14646388304e1f ipsum sit dolor")
   ["37i9dQZF1DZ06evO2HVeMk"])
  (is
   (extract-albums "https://open.spotify.com/album/2exrjacqbNx8HAQLfn96qJ?si=XIuiXtuQQcSk3O-18F3mPw")
   ["spotify:album:2exrjacqbNx8HAQLfn96qJ"]))


test/r_tae/music_is_midbot_test.clj
(ns r-tae.music-is-midbot-test
  (:require [clojure.test :refer :all]
            [r-tae.music-is-midbot :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


build.clj
(ns build
  (:refer-clojure :exclude [test])
  (:require [clojure.tools.deps :as t]
            [clojure.tools.build.api :as b]))

(def lib 'mim.bot)
(def version "0.1.3")
(def main 'mim.bot)
(def class-dir "target/classes")

(defn test "Run all the tests." [opts]
  (println "\nRunning tests...")
  (let [basis    (b/create-basis {:aliases [:test]})
        combined (t/combine-aliases basis [:test])
        cmds     (b/java-command
                  {:basis basis
                   :java-opts (:jvm-opts combined)
                   :main      'clojure.main
                   :main-args ["-m" "cognitect.test-runner"]})
        {:keys [exit]} (b/process cmds)]
    (when-not (zero? exit) (throw (ex-info "Tests failed" {}))))
  opts)

(defn- uber-opts [opts]
  (assoc opts
         :lib lib :main main
         :uber-file (format "target/%s-%s.jar" lib version)
         :basis (b/create-basis {})
         :class-dir class-dir
         :src-dirs ["src"]
         :ns-compile [main]))

(defn ci "Run the CI pipeline of tests (and build the uberjar)." [opts]
  (test opts)
  (b/delete {:path "target"})
  (let [opts (uber-opts opts)]
    (println "\nCopying source...")
    (b/copy-dir {:src-dirs ["resources" "src"] :target-dir class-dir})
    (println (str "\nCompiling " main "..."))
    (b/compile-clj opts)
    (println "\nBuilding JAR...")
    (b/uber opts))
  opts)


src/mim/bot.clj
(ns mim.bot
  (:require
   [mim.spotify :as spotify]
   [mim.state :refer [get-state update-state]]
   [mim.util :refer [generate-token]]
   [clojure.set :as set]
   [discljord.connections :as con]
   [discljord.messaging :as rest]
   [discljord.events :as events]
   [clojure.core.async :as a]
   [dotenv :refer [env]]
   [slash.command :as cmd]
   [slash.core :as slash]
   [slash.command.structure :as stc]
   [slash.gateway :refer [gateway-defaults]]
   [com.brunobonacci.mulog :as u]
   [slash.response :as rsp :refer [channel-message ephemeral]])
  (:gen-class))

(def discord-token
  (env :DISCORD_AUTH_TOKEN))
(def application-id "1141359842955624490")

(u/start-publisher! {:type :console})
(def rest-conn (rest/start-connection! discord-token))

(rest/bulk-overwrite-global-application-commands!
 rest-conn application-id
 [(stc/command
   "echo"
   "Echoes your input"
   :options
   [(stc/option "input"
                "Your input" :string :required true)])
  (stc/command
   "status"
   "Check playlist bot status"
   :options [])
  (stc/command
   "auth"
   "Fetch Spotify auth URL"
   :options [])
  (stc/command
   "playlist"
   "Create or use an existing Spotify playlist"
   :options
   [(stc/option "url"
                "Use an existing playlist" :string :required false)
    (stc/option "name"
                "Make a new playlist" :string :required false)])])

(defn respond-interaction [id token {:keys [type data]}]
  (rest/create-interaction-response! rest-conn id token type :data data))


(defn if-mod
  [member mod-branch not-mod-branch]
  (if  (seq (set/intersection (set (:roles member))
                              #{"1143560732089262090" ;;testing server
                                "1079637398281011200" ;; music-is-mid
                                }))
    (mod-branch) (not-mod-branch)))


;; (cmd/defhandler echo-handler
;;   ["echo"] ; Command path
;;   {:keys [id token] :as _interaction} ; Interaction binding
;;   [input] ; Command options
;;   (->> (channel-message {:content input})
;;        ephemeral
;;        (respond-interaction id token)))
(cmd/defhandler status-handler
  ["status"] ; Command path
  {:keys [id token] :as interaction} ; Interaction binding
  [] ; Command options
  (u/log ::handle-status-command)
  (u/log ::interaction :data interaction)
  (->> (channel-message {:content (str "Status:\nSpotify account: " (:display_name (spotify/get-user))
                                       "\nPlaylist: " (get-state :playlist :external_urls :spotify))})
       ephemeral
       (respond-interaction id token)))
(cmd/defhandler auth-handler
  ["auth"]
  {:keys [id token member] :as _interaction}
  []
  (if-mod member
          #(do (u/log ::handle-auth-command :mod true)
               (update-state [:auth :csrf-token] (generate-token))
               (->> (channel-message {:content
                                      (str "Follow link to authenticate your spotify account: " (env :BASE_URL) "/interaction")})
                    ephemeral
                    (respond-interaction id token)))
          #(do (u/log ::handle-auth-command :mod false)
               (->> (channel-message {:content
                                      (str "You must be a mod to run this command.")})
                    ephemeral
                    (respond-interaction id token)))))


(cmd/defhandler playlist-handler
  ["playlist"]
  {:keys [id token member] :as _interaction}
  [url name]
  (if-mod member
          #(do (u/log ::handle-playlist-command :mod true :url url :name name)
               (->>
                (cond
                  url (do (spotify/set-playlist url) (channel-message {:content (str "Using " (get-state :playlist :external_urls :spotify))}))
                  name (do (spotify/new-playlist name) (channel-message {:content (str "Created a new playlist: " (get-state :playlist :external_urls :spotify))}))
                  :else (channel-message {:content "Please give a URL or name for your playlist"}))
                ephemeral
                (respond-interaction id token)))
          #(do (u/log ::handle-playlist-command :mod false)
               (->> (channel-message {:content
                                      (str "You must be a mod to run this command.")})
                    ephemeral
                    (respond-interaction id token)))))

(cmd/defpaths command-paths
  status-handler
  auth-handler
  playlist-handler)

(defn handle-ready [_ _data] (u/log ::connected))

(defn handle-message-create
  [_ {{bot :bot} :author :keys [content] :as _data}]
  (when-not bot
    (let [tracks (spotify/filter-existing-tracks (spotify/extract-tracks content))]
      (when-not (empty? tracks)
        (u/log ::add-to-playlist :tracks tracks)
        (spotify/add-to-playlist tracks)))))

(def event-handlers
  {:ready [#'handle-ready]
   :interaction-create
   [#(slash/route-interaction (assoc gateway-defaults
                                     :application-command command-paths) %2)]
   :message-create [#'handle-message-create]})

(defn -main
  [& _args]
  (spotify/start-auth-server)
  (let [channel (a/chan 100)
        conn-ch (con/connect-bot! discord-token channel
                                  :intents #{:guilds :guild-messages})]
    (events/message-pump! channel
                          (partial events/dispatch-handlers event-handlers))
    (rest/stop-connection! rest-conn)
    (con/disconnect-bot! conn-ch)
    (a/close! channel)))


.clj-kondo/hooks/slash.clj
(ns hooks.slash
  (:require [clj-kondo.hooks-api :as api]))

(defn strip-pattern [pattern]
  (api/vector-node (vec (remove api/string-node? (:children pattern)))))

(defn transform-handler-contents [[pattern interact-sym opt-sym & body]]
  (api/list-node
   (list*
    (api/token-node 'let)
    (api/vector-node
     [(strip-pattern pattern) (api/token-node nil)
      interact-sym (api/token-node nil)
      opt-sym (api/token-node nil)])
    body)))

(defn handler [{:keys [node]}]
  {:node (transform-handler-contents (rest (:children node)))})

(defn defhandler [{:keys [node]}]
  (let [[name & handler-contents] (rest (:children node))]
    {:node (api/list-node (list (api/token-node 'def) name (transform-handler-contents handler-contents)))}))

(defn group [{:keys [node]}]
  (let [[pattern & handlers] (rest (:children node))]
    {:node (api/list-node
            (list*
             (api/token-node 'let)
             (api/vector-node [(strip-pattern pattern) (api/token-node nil)])
             handlers))}))

(defn defpaths [{:keys [node]}]
  (let [[name & handlers] (rest (:children node))]
    {:node (api/list-node
            (list
             (api/token-node 'def)
             name
             (api/vector-node (vec handlers))))}))

src/r_tae/music_is_midbot.clj
(require '[discljord.connections :as c])
(require '[discljord.messaging :as m])
(require '[discljord.events :as e])
(require '[clojure.core.async :as a])
(require '[loudmoauth.core :as lm])
(require '[ring.adapter.jetty :as ringj])
(require '[ring.util.response :as ringr])
(require '[ring.middleware.params :as ringp])
(require '[ring.middleware.keyword-params :as ringkp])
(require '[clj-spotify.core :as sptfy])
(require '[clj-spotify.util :as util])
(require '[clojure.string :as str])
(require '[clojure.core.match :refer [match]])
(require '[dotenv :refer [env]])

(def state (atom nil))
(def auth-message
  (str "Follow link to authenticate your spotify account: " (env :BASE_URL) "/interact"))
(def token
  (env :DISCORD_AUTH_TOKEN))
(def spotify-client-id (env :SPOTIFY_OAUTH2_CLIENT_ID))
(def spotify-client-secret (env :SPOTIFY_OAUTH2_CLIENT_SECRET))
(def intents #{:guilds :guild-messages})

;; TODO: handle albums
(defn extract-songs [msg]
  (map (fn [match]
         (clojure.string/join ":" (cons "spotify" (rest match))))
       (re-seq #"https://open.spotify.com/(track)/([-a-zA-Z0-9]*)(?:\?.*)?" msg)))

(defn has-spotify-tracks? [msg] (re-seq #"https://open.spotify.com/(track)/([-a-zA-Z0-9]*)(?:\?.*)?" msg))

(defn create-playlist []
  (sptfy/create-a-playlist {:user_id (get (sptfy/get-current-users-profile {} (lm/oauth-token :spotify)) :id) :name
                            "Music is mid recs" :public true} (lm/oauth-token :spotify)))

;; TODO: don't hardcode the playlist ID
(defn add-to-playlist [uris]
  (match [(sptfy/add-tracks-to-a-playlist
           {:user_id (get (sptfy/get-current-users-profile {} (lm/oauth-token :spotify)) :id)
            :playlist_id (get @state :playlist_id) :uris uris}
           (lm/oauth-token :spotify))]
    [{:snapshot_id _}] "Added to playlist"
    :else "Error"))

(defn handler [request]
  (condp = (:uri request)
    "/oauth2" (lm/parse-params request)
    "/interact"  (ringr/redirect (lm/user-interaction))
    {:status 200
     :body (:uri request)}))

(defn run-auth-server
  "Starts our test web server."
  []
  (future (ringj/run-jetty (ringp/wrap-params (ringkp/wrap-keyword-params handler))  {:port 3000}))
  (lm/add-provider {:base-url "https://accounts.spotify.com"
                    :auth-endpoint "/authorize"
                    :token-endpoint "/api/token"
                    :client-id spotify-client-id
                    :redirect-uri (str (env :BASE_URL) "/oauth2")
                    :scope "playlist-read-private playlist-modify-public playlist-modify-private ugc-image-upload"
                    :custom-query-params {:show-dialog "true"}
                    :client-secret spotify-client-secret
                    :provider :spotify}))

(run-auth-server)

(defmulti handle-event
  (fn [event-type _event-data]
    event-type))

(defmethod handle-event :default
  [_event-type _event-data])
(defn handle-new-playlist [channel-id]
  (let [playlist (create-playlist)]
    (swap! state assoc :playlist_id (:id playlist))
    (m/create-message! (:messaging @state) channel-id :content
                       (str "<" (get-in playlist [:external_urls :spotify]) ">"))))

;; TODO: error handling
;; TODO: limit to one channel
(defn set-playlist? [a] (str/starts-with? a "!set-playlist"))
(defmethod handle-event :message-create
  [_event-type {{bot :bot} :author :keys [channel-id content id]}]
  (when-not bot (match [content]
                  ["!disconnect"] (c/disconnect-bot! (:connection @state))
                  ["!auth"] (m/create-message! (:messaging @state) channel-id :content auth-message)
                  ;; TODO: register commands properly
                  [_ :guard set-playlist?] (swap! state assoc :playlist_id
                                                  (nth (re-find #"https://open.spotify.com/playlist/(\w*)" content) 1))
                  ;; TODO: proper error handling and replies
                  ["!new-playlist"] (handle-new-playlist channel-id)
                  :else (if (has-spotify-tracks? content)
                          (m/create-message! (:messaging @state) channel-id :content (add-to-playlist (extract-songs content)))
                          ;; TODO: don't error if we can't do anything with the message
                          (m/create-reaction! (:messaging @state) channel-id id "🆘")))))

(a/go
  (let [event-ch (a/chan 100)
        connection-ch (c/connect-bot! token event-ch :intents intents)
        messaging-ch (m/start-connection! token)
        init-state {:connection connection-ch
                    :event event-ch
                    :messaging messaging-ch}]
    (reset! state init-state)
    (try (e/message-pump! event-ch handle-event)
         (finally
           (m/stop-connection! messaging-ch)
           (a/close!           event-ch)))))


src/mim/spotify.clj
(ns mim.spotify
  (:require [clj-http.client :as http]
            [clj-http.util :refer [url-encode]]
            [clj-spotify.core :as sptfy]
            [com.brunobonacci.mulog :as u]
            [dotenv :refer [env]]
            [mim.state :refer [get-state update-state]]
            [ring.adapter.jetty :as ringj]
            [ring.util.response :as ringr]
            [ring.middleware.params :as ringp]
            [ring.middleware.keyword-params :as ringkp]
            [slingshot.slingshot :refer [try+]]))

(def ^:private track-re #"https://open.spotify.com/(track)/(\w*)")
(def ^:private playlist-re #"https://open.spotify.com/(playlist)/(\w*)")
(def ^:private album-re #"https://open.spotify.com/(album)/(\w*)")

(declare refresh-tokens)
;; (defn refresh-token [] (auth/refresh-token :spotify))
(defn log-out [] (update-state :auth nil))

; TODO: somehow reply on persistant error
(defn- retry-on-error
  "Retry after refreshing the spotify token if `call` evaluates to `{:error _}`
       (retry-on-error '(sptfy/get-current-users-profile {} (lm/oauth-token :spotify))"
  [call]
  (let [result (call)]
    (if (:error result)
      (do (u/log ::retrying :call call) (refresh-tokens (get-state :auth :access_token))
          (call))
      result)))

(defn extract-tracks
  "Returns the track IDs of any spotify URLs in `msg`
   ```clojure
      (extract-tracks \"https://open.spotify.com/track/7oJgLeTIETiDEqRFaWbS3k?si=b606403c7cb34870\")
      => [\"spotify:track:7oJgLeTIETiDEqRFaWbS3k\"]
   ```"
  [message]
  (map (fn [match]
         (clojure.string/join ":" (cons "spotify" (rest match))))
       (re-seq #"https://open.spotify.com/(track)/([-a-zA-Z0-9]*)(?:\?.*)?" message)))
(defn extract-playlists
  "Returns the playlist IDs of any spotify URLs in `msg`
      (extract-playlists \"https://open.spotify.com/playlist/37i9dQZF1DZ06evO2HVeMk?si=5f14646388304e1f\")
      => [\"37i9dQZF1DZ06evO2HVeMk\"]
   Note this lacks the 'spotify:' prefix because the Spotify API don't seem to expect it anywhere."
  [message]
  (re-find playlist-re message))
(defn extract-albums
  "Returns the album IDs of any spotify URLs in `msg`
      (extract-albums \"https://open.spotify.com/album/2exrjacqbNx8HAQLfn96qJ?si=XIuiXtuQQcSk3O-18F3mPw\")
      => [\"spotify:album:2exrjacqbNx8HAQLfn96qJ\"]"
  [message]
  (re-find album-re message))

(declare get-fresh-token)

(defn filter-existing-tracks [tracks]
  (let [playlist (sptfy/get-a-playlist {:playlist_id (get-state :playlist :id)} (get-fresh-token))
        existing-tracks (map (comp :uri :track) (:items (:tracks playlist)))]
    (filter #(not (some (fn [t] (= t %))
                        existing-tracks))  tracks)))

(defn get-user []
  (sptfy/get-current-users-profile {} (get-fresh-token)))

(defn set-playlist [msg]
  (let [playlist-id (second (extract-playlists msg))]
    (update-state [:playlist]
                  (sptfy/get-a-playlist {:playlist_id playlist-id} (get-fresh-token)))))

(defn new-playlist [name]
  (let [token (get-fresh-token)]
    (update-state [:playlist]
                  (sptfy/create-a-playlist {:user_id (:id (sptfy/get-current-users-profile {} token))
                                            :name name} token))))

(defn add-to-playlist [tracks]
  (sptfy/add-tracks-to-a-playlist
   {;; :user_id (get (sptfy/get-current-users-profile {} (lm/oauth-token :spotify)) :id)
    :playlist_id (get-state :playlist :id) :uris tracks}
   (get-fresh-token)))

(def ^:private oauth2-params
  {:base-url "https://accounts.spotify.com"
   :auth-endpoint "/authorize"
   :token-endpoint "/api/token"
   :client-id (env :SPOTIFY_OAUTH2_CLIENT_ID)
   :redirect-uri (str (env :BASE_URL) "/oauth2")
   :scope "playlist-read-private playlist-modify-public playlist-modify-private ugc-image-upload"
   :show-dialog "true"
   :client-secret (env :SPOTIFY_OAUTH2_CLIENT_SECRET)})

(defn authorize-uri [client-params csrf-token]
  (str
   (:base-url client-params)
   (:auth-endpoint client-params)
   "?response_type=code"
   "&client_id="
   (url-encode (:client-id client-params))
   "&redirect_uri="
   (url-encode (:redirect-uri client-params))
   "&scope="
   (url-encode (:scope client-params))
   "&state="
   (url-encode csrf-token)))

(defn get-authentication-response [csrf-token response-params]
  (if (= csrf-token (:state response-params))
    (try
      (-> (http/post (str (:base-url oauth2-params) (:token-endpoint oauth2-params))
                     {:form-params {:code         (:code response-params)
                                    :grant_type   "authorization_code"
                                    :client_id    (:client-id oauth2-params)
                                    :redirect_uri (:redirect-uri oauth2-params)}
                      :basic-auth [(:client-id oauth2-params) (:client-secret oauth2-params)]
                      :as          :json})
          :body)
      (catch Exception e (u/log ::exception :exception e)))
    nil))

(defn- refresh-tokens
  "Request a new token pair"
  [refresh-token]
  (try+
   (let [{{access-token :access_token refresh-token :refresh_token} :body}
         (http/post (str (:base-url oauth2-params) (:token-endpoint oauth2-params))
                    {:form-params {:grant_type       "refresh_token"
                                   :refresh_token    refresh-token}
                     :basic-auth [(:client-id oauth2-params) (:client-secret oauth2-params)]
                     :as          :json})]
     (when (and access-token refresh-token) (update-state [:auth :tokens] [access-token refresh-token]))
     access-token)
   (catch [:status 401] _ nil)))

(defn get-fresh-token
  "Returns current token pair if they have not expired, or a refreshed token pair otherwise"
  []
  (let [[access-token refresh-token] (get-state :auth :tokens)
        test-response (sptfy/get-current-users-profile {} access-token)]
    (if (contains? test-response :error)
      (refresh-tokens refresh-token)
      access-token)))

(defn auth-handler [request]
  (condp = (:uri request)
    "/oauth2" (let [{access-token :access_token refresh-token :refresh_token} (get-authentication-response (get-state :auth :csrf-token) (:params request))]
                (when (and access-token refresh-token)
                  (u/log ::oauth-flow-completed :request request)
                  (update-state [:auth :tokens] [access-token refresh-token]))
                {:status 200 :body "<script>window.close()</script>"})
    "/interaction" (ringr/redirect (authorize-uri oauth2-params (get-state :auth :csrf-token)))
    {:status 200
     :body (:uri request)}))

(defn start-auth-server
  "Starts our auth web server."
  []
  (future (-> auth-handler
              ringkp/wrap-keyword-params
              ringp/wrap-params
              (ringj/run-jetty {:port 3000}))))

;; hit api
;; if it errors with 401
;;     - refresh token
;;     - try again
;;     - if it fails again, DM me
;; if it errors with 403, don't bother trying to recover, DM me
;; if it errors with 439, wait 2 seconds, if it fails after that then DM me


;; https://open.spotify.com/track/61NudzDPuUUH9NuhKMDgRD?si=ec51a9fa907b4ed0
;; https://open.spotify.com/track/2aQ8NyRJcsQFgb8jwUil90?si=c0671cf6eeb04970")

